File: packages/core/src/validator.ts
import Ajv, { ErrorObject } from "ajv/dist/2020.js";
import addFormats from "ajv-formats";
import schema from "../schema/v1.2.json";

const ajv = new Ajv({ allErrors: true, strict: false });
addFormats(ajv);

const validate = ajv.compile(schema as object);

export interface ValidationResult {
  valid: boolean;
  errors?: ErrorObject[] | null;
  deployable?: boolean;
  missingRequirements?: string[];
}

export function validateArtifact(artifact: unknown): ValidationResult {
  const valid = validate(artifact) as boolean;
  const obj = artifact as { deployable?: boolean; missing_requirements?: string[] };

  return {
    valid,
    errors: validate.errors ?? undefined,
    deployable: obj?.deployable,
    missingRequirements: obj?.missing_requirements,
  };
}

File: packages/langchain-python/safestop_langchain/validator.py
import json
from pathlib import Path

from jsonschema import validate as jsonschema_validate, ValidationError

SCHEMA_PATH = Path(__file__).resolve().parent / "schema" / "v1.2.json"


def validate_artifact(artifact: dict) -> dict:
    """Validate a Safestop artifact against the v1.2 schema.

    Returns a dict with:
        - valid: bool
        - deployable: bool (from artifact)
        - missing_requirements: list (when deployable is False)
        - errors: list (when valid is False)
    """
    with open(SCHEMA_PATH, encoding="utf-8") as f:
        schema = json.load(f)

    try:
        jsonschema_validate(instance=artifact, schema=schema)
        return {
            "valid": True,
            "deployable": artifact.get("deployable", False),
            "missing_requirements": artifact.get("missing_requirements", []),
        }
    except ValidationError as e:
        return {
            "valid": False,
            "errors": [e.message],
        }
